pallets/foreign-investments/src/impls/invest.rs

```diff
@@ -0,0 +1,967 @@
+	/// this can never occur!
+	fn handle_increase_non_foreign(
+		&self,
+		swap: Swap<Balance, Currency>,
+	) -> Result<Self, DispatchError> {
+		match &self {
+			Self::NoState => Ok(Self::InvestmentOngoing {
+				invest_amount: swap.amount,
+			}),
+			Self::InvestmentOngoing { invest_amount } => Ok(Self::InvestmentOngoing {
+				invest_amount: invest_amount.ensure_add(swap.amount)?,
+			}),
+			Self::ActiveSwapIntoPoolCurrency { .. }
+			| Self::ActiveSwapIntoForeignCurrency { .. }
+			| Self::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing { .. }
+			| Self::ActiveSwapIntoForeignCurrencyAndInvestmentOngoing { .. }
+			| Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone { .. }
+			| Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+				..
+			} => Err(DispatchError::Other(
+				"Invalid invest state when transitioning an increased swap order with the same in- \
+				 and outgoing currency",
+			)),
+			_ => Err(DispatchError::Other(
+				"Invalid invest state, should automatically be transitioned into state without \
+				 AndSwapIntoForeignDone",
+			)),
+		}
+	}
+
+	/// Handle decrease transitions for the same incoming and outgoing
+	/// currencies.
+	///
+	/// NOTE: We can ignore handling all states which include
+	/// `SwapIntoForeignDone` without `ActiveSwapIntoForeignCurrency` as we
+	/// consume the done amount and transition in the post transition phase.
+	/// Moreover, we can ignore any state which involves an active swap, i.e.
+	/// `ActiveSwapInto{Pool, Return}Currency`, as these must not exist if the
+	/// in and out currency is the same.
+	/// To be safe and to not make any unhandled assumptions, we throw
+	/// `DispatchError::Other` for these states though we need to make sure
+	/// this can never occur!
+	fn handle_decrease_non_foreign(
+		&self,
+		swap: Swap<Balance, Currency>,
+	) -> Result<Self, DispatchError> {
+		if let Self::InvestmentOngoing { invest_amount } = &self {
+			if swap.amount < *invest_amount {
+				Ok(InvestState::SwapIntoForeignDoneAndInvestmentOngoing {
+					done_swap: swap,
+					invest_amount: invest_amount.ensure_sub(swap.amount)?,
+				})
+			} else {
+				Ok(Self::SwapIntoForeignDone { done_swap: swap })
+			}
+		}
+		// should never occur but let's be safe here
+		else {
+			Err(DispatchError::Other(
+				"Invalid invest state when transitioning a decreased swap order with the same in- \
+				 and outgoing currency",
+			))
+		}
+	}
+
+	/// Update or kill the state's unprocessed investing amount.
+	/// * If the state includes `InvestmentOngoing`, either update or remove the
+	///   invested amount.
+	/// * Else the unprocessed amount should be zero. If it is not, state is
+	///   corrupted as this reflects the investment was increased improperly.
+	fn handle_collect(&self, unprocessed_amount: Balance) -> Result<Self, DispatchError> {
+		match *self {
+			Self::InvestmentOngoing { .. } => {
+				if unprocessed_amount.is_zero() {
+					Ok(Self::NoState)
```

Comment:
If this state exists, I would remove its docs saying that it is never stored on chain (this comes probably from the refactor removing the `None` I guess)